### ISSUE TITLE
failure to save a new_record? has_one association is silently swallowed when autosave is not specified

### ISSUE THREAD

**Issue by bughit** [rails/rails#21594]
``` ruby
      def save_has_one_association(reflection)
        association = association_instance_get(reflection.name)
        record      = association && association.load_target

        if record && !record.destroyed?
          autosave = reflection.options[:autosave]

          if autosave && record.marked_for_destruction?
            record.destroy
          elsif autosave != false
            key = reflection.options[:primary_key] ? send(reflection.options[:primary_key]) : id

            if (autosave && record.changed_for_autosave?) || new_record? || record_changed?(reflection, record, key)
              unless reflection.through_reflection
                record[reflection.foreign_key] = key
              end

              saved = record.save(:validate => !autosave)
              raise ActiveRecord::Rollback if !saved && autosave
              saved
            end
          end
        end
      end
```

`raise ActiveRecord::Rollback if !saved && autosave`

**Comment by rafaelfranca** [comment]
Could you please use the reproduction script so we can know what you
expect?

**Comment by bughit** [comment]
Since new associations are supposed to be saved, per the docs, failure should raise and abort, just as when autosave is true

```ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  # Activate the gem you are reporting the issue against.
  gem 'activerecord', '4.2.4'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_one :comment
end

class Comment < ActiveRecord::Base
  belongs_to :post
  before_save {false}
end

class BugTest < Minitest::Test
  def test_association_stuff
    post = Post.new
    comment = post.build_comment
    post_saved = post.save

    refute comment.persisted?
    refute post_saved

  end
end
```

**Comment by bughit** [comment]
@rafaelfranca Is this is a bug?

**Comment by rafaelfranca** [comment]
I don't know. I didn't have time to investigate. Please do investigate

**Comment by bughit** [comment]
I already showed why it's happening in the first post

`raise ActiveRecord::Rollback if !saved && autosave`

it only raises if the autosave option is present

I'm just asking if this is a bug, as it seems to me, or if there's a reason for this?

**Comment by rafaelfranca** [comment]
I think it is a bug. Could you send a PR?

**Comment by sgrif** [comment]
Worth noting that this is pretty explicitly intended, and we have tests for this behavior. https://github.com/rails/rails/blob/master/activerecord/test/cases/autosave_association_test.rb#L97
https://github.com/rails/rails/commit/da3c21ead59cb47b8f4c69c6bd95f225a9c8b479

That said, the original reason for the change was to resolve https://rails.lighthouseapp.com/projects/8994/tickets/2249, and I think it would be fine to fail to save the parent. This would be a pretty significant change in behavior though, and I think we should go through a deprecation cycle. This was the intentionally added behavior, and it's been around for too long to just call it a bug.

**Comment by bughit** [comment]
related problems (when autosave is set): #22183, #22184

we have the following problems with the saving of associations:

1. without autosave, fails to abort on association failure leaving the object graph in an inconsistent state
2. with autosave: aborts but pretends success (save! does not raise, persisted? is true)

so one can not depend on saving of associations at all

**Comment by bughit** [comment]
@sgrif 

>  I think we should go through a deprecation cycle

It does not make sense to protect something that is this broken, I think it would be much better to fix all three of these bugs as soon as possible without any deprecation cycles.  It would then simply work as the docs say it should, without any surprises.

**Comment by bughit** [comment]
@rafaelfranca @sgrif

Am I misunderstanding the severity of this set of bugs?

* without `autosave: true`, fails to abort on association failure 
* with `autosave: true` aborts on association failure but pretends success (save! does not raise, persisted? is true)

in both cases the expected saves have not taken place when success is reported.  Does this not make the feature of "association saving" too unreliable to use?

I proposed fixes for #22183  #22184, can someone please review and comment on them.  And for this one, do we really need a deprecation cycle?  Who could possibly be relying on the current behavior of the associations silently not saving when you expect them to?

**Comment by caifara** [comment]
This is still an issue on rails 7.1, it's also wildly unexpected and breaks the elegance of having a single save call and thus depending on Rails to save a tree of objects.
